# Lesson 01 ... Welcome to pandas

We will start with importing some libraries we need and then play with some data to understand basic python commands. What data shall we work with? Well, let us pull down some data on criminal incidences that were reported.

First we install a particular library called `pandas` and in the command that follows, note that `pd` is just the alias that pandas assumes so that we can type `pd` and have all the `pandas` commands at our disposal.

In [4]:
import pandas as pd 
import numpy as np

The crime incident reports data are [available here]("https://data.boston.gov/dataset/6220d948-eae2-4e4b-8723-2dc8e67722a3/resource/12cb3883-56f5-47de-afa5-3b1cf61b257b/download/tmpayw7hysb.csv") and span multiple years so we may end up working only with 2019 data but for now we proceed by gathering everything.

In the command below, the key part is `pd.read_csv()` and inside it is the URL for the comma-separated variable file. Once the file is downloaded by `pandas` we are saving it in python with the name `df` 

Note that data-sets, data-files are usually referred to as a `data-frame` in python and hence the alias of `df`.


In [5]:
df = pd.read_csv("https://data.boston.gov/dataset/6220d948-eae2-4e4b-8723-2dc8e67722a3/resource/12cb3883-56f5-47de-afa5-3b1cf61b257b/download/tmpayw7hysb.csv")

/srv/conda/envs/notebook/lib/python3.7/site-packages/IPython/core/interactiveshell.py:3072: DtypeWarning: Columns (0,2,6,12) have mixed types.Specify dtype option on import or set low_memory=False.
  interactivity=interactivity, compiler=compiler, result=result)


Let us look at the first 5 rows of data to get a feel for the layout. The command is `.head(5)`

In [7]:
df.head(5)

INCIDENT_NUMBER  OFFENSE_CODE OFFENSE_CODE_GROUP        OFFENSE_DESCRIPTION  \
0       TESTTEST2           423                NaN       ASSAULT - AGGRAVATED   
1       S97333701          3301                NaN             VERBAL DISPUTE   
2       S47513131          2647                NaN  THREATS TO DO BODILY HARM   
3       I92102201          3301                NaN             VERBAL DISPUTE   
4       I92097173          3115                NaN         INVESTIGATE PERSON   

   DISTRICT REPORTING_AREA SHOOTING     OCCURRED_ON_DATE  YEAR  MONTH  \
0  External                       0  2019-10-16 00:00:00  2019     10   
1        C6            915        0  2020-07-18 14:34:00  2020      7   
2       E18            530        0  2020-06-24 10:15:00  2020      6   
3       E13            583        0  2019-12-20 03:08:00  2019     12   
4       C11            355        0  2019-10-23 00:00:00  2019     10   

  DAY_OF_WEEK  HOUR UCR_PART          STREET        Lat       Long  \
0   Wednesday     0      NaN    RIVERVIEW DR        NaN        NaN   
1    Saturday    14      NaN  MARY BOYLE WAY  42.330813 -71.051368   
2   Wednesday    10      NaN    READVILLE ST  42.239491 -71.135954   
3      Friday     3      NaN          DAY ST  42.325122 -71.107779   
4   Wednesday     0      NaN       GIBSON ST  42.297555 -71.059709   

                      Location  
0     (0.00000000, 0.00000000)  
1  (42.33081300, -71.05136800)  
2  (42.23949100, -71.13595400)  
3  (42.32512200, -71.10777900)  
4  (42.29755500, -71.05970900)

What about the last 10 rows of the data?

In [8]:
df.tail(10)

INCIDENT_NUMBER  OFFENSE_CODE OFFENSE_CODE_GROUP  \
515072       102095489          3115                NaN   
515073       102091671          2647                NaN   
515074       020224065          3018                NaN   
515075       020202856          2672                NaN   
515076       020063425          3114                NaN   
515077       020062356          3115                NaN   
515078       020054040          3501                NaN   
515079       020046400          1501                NaN   
515080       020038446          1501                NaN   
515081       020030892           540                NaN   

                                      OFFENSE_DESCRIPTION DISTRICT  \
515072                                 INVESTIGATE PERSON      E18   
515073                          THREATS TO DO BODILY HARM       B3   
515074                      SICK/INJURED/MEDICAL - POLICE       B2   
515075                                 BIOLOGICAL THREATS       B2   
515076                               INVESTIGATE PROPERTY       A7   
515077                                 INVESTIGATE PERSON      E18   
515078                                     MISSING PERSON      C11   
515079  WEAPON VIOLATION - CARRY/ POSSESSING/ SALE/ TR...       B2   
515080  WEAPON VIOLATION - CARRY/ POSSESSING/ SALE/ TR...       B2   
515081                              BURGLARY - COMMERICAL      C11   

       REPORTING_AREA SHOOTING     OCCURRED_ON_DATE  YEAR  MONTH DAY_OF_WEEK  \
515072            520        0  2019-11-25 16:30:00  2019     11      Monday   
515073            417        0  2019-11-12 12:00:00  2019     11     Tuesday   
515074            282        0  2020-03-19 07:30:00  2020      3    Thursday   
515075            282        0  2020-03-19 08:30:00  2020      3    Thursday   
515076             21        0  2020-09-01 00:00:00  2020      9     Tuesday   
515077            520        0  2020-08-28 18:39:00  2020      8      Friday   
515078                       0  2020-07-30 15:30:00  2020      7    Thursday   
515079            330        0  2020-07-02 01:38:00  2020      7    Thursday   
515080            300        0  2020-06-03 01:15:00  2020      6   Wednesday   
515081            380        0  2020-05-03 00:00:00  2020      5      Sunday   

        HOUR UCR_PART         STREET        Lat       Long  \
515072    16      NaN  HYDE PARK AVE  42.256215 -71.124019   
515073    12      NaN        MORA ST  42.282081 -71.073648   
515074     7      NaN  WASHINGTON ST  42.353272 -71.173738   
515075     8      NaN      WARREN ST  42.328234 -71.083289   
515076     0      NaN       PARIS ST  42.374426 -71.035278   
515077    18      NaN  HYDE PARK AVE  42.256215 -71.124019   
515078    15      NaN      GIBSON ST  42.297555 -71.059709   
515079     1      NaN    PASADENA RD  42.305760 -71.083771   
515080     1      NaN  WASHINGTON ST  42.323807 -71.089150   
515081     0      NaN  GALLIVAN BLVD  42.283700 -71.047761   

                           Location  
515072  (42.25621500, -71.12401900)  
515073  (42.28208100, -71.07364800)  
515074  (42.35327200, -71.17373800)  
515075  (42.32823400, -71.08328900)  
515076  (42.37442600, -71.03527800)  
515077  (42.25621500, -71.12401900)  
515078  (42.29755500, -71.05970900)  
515079  (42.30576000, -71.08377100)  
515080  (42.32380700, -71.08915000)  
515081  (42.28370000, -71.04776100)

Let us look at the contents of the data-frame ... 

| Column Name | Description |
| :--          | :--- |
| [incident_num] [varchar](20) NOT NULL, | Internal BPD report number |
| [offense_code][varchar](25) NULL,| Numerical code of offense description |
| [Offense_Code_Group_Description][varchar](80) NULL, | Internal categorization of [offense_description] |
| [Offense_Description][varchar](80) NULL, | Primary descriptor of incident |
| [district] [varchar](10) NULL,| What district the crime was reported in |
| [reporting_area] [varchar](10) NULL, | RA number associated with the where the crime was reported from. |
| [shooting][char] (1) NULL, | Indicated a shooting took place. |
| [occurred_on] [datetime2](7) NULL, | Earliest date and time the incident could have taken place |
| [UCR_Part] [varchar](25) NULL,| Universal Crime Reporting Part number (1,2, 3) |
| [street] [varchar](50) NULL,| Street name the incident took place |


Offense Codes are [available here](https://data.boston.gov/dataset/6220d948-eae2-4e4b-8723-2dc8e67722a3/resource/3aeccf51-a231-4555-ba21-74572b4c33d6/download/rmsoffensecodes.xlsx)

We could also look at the offense codes by reading them in as a data-frame. This is an Excel file so we will have to switch to `.read_excel()`


In [9]:
offense_codes = pd.read_excel("https://data.boston.gov/dataset/6220d948-eae2-4e4b-8723-2dc8e67722a3/resource/3aeccf51-a231-4555-ba21-74572b4c33d6/download/rmsoffensecodes.xlsx")

In [10]:
print(offense_codes)

     CODE                                       NAME
0     612           LARCENY PURSE SNATCH - NO FORCE 
1     613                        LARCENY SHOPLIFTING
2     615    LARCENY THEFT OF MV PARTS & ACCESSORIES
3    1731                                     INCEST
4    3111                  LICENSE PREMISE VIOLATION
..    ...                                        ...
571  1806  DRUGS - CLASS B TRAFFICKING OVER 18 GRAMS
572  1807  DRUGS - CLASS D TRAFFICKING OVER 50 GRAMS
573  1610    HUMAN TRAFFICKING - COMMERCIAL SEX ACTS
574  2010                              HOME INVASION
575  1620  HUMAN TRAFFICKING - INVOLUNTARY SERVITUDE

[576 rows x 2 columns]


The next step would be to see how many data points we have, and what the minimum, maximum values, what is the average, etc. This can be done with `.describe()`

In [ ]:
df.describe()

By default the command will report the values with decimals but we may not want that. Decimals can be `rounded` or removed altogether as shown below.

In [ ]:
df.describe().round(2)

In [ ]:
df.describe().round(0)

Note a few things here. 

* We have a total of 515082 incidents logged. But the latitude and longitude are availale for no more than 485909 incidents. 


Say we want to restrict the dataframe just to 2020. How can we do that?

In [ ]:
df20 = df[ df['YEAR'] == 2020 ]

Notice the sequence here `dataframe[ dataframe['column-name'] == somevalue ]` & pay attention to the double equal sign `==` which is a strict equality. 

In [ ]:
df20.describe()

At this point we might be curious to know what types of offenses are most often reported? Before we that, however, let us also see how many unique values of OFFENSE_CODE are there

In [ ]:
df20['OFFENSE_CODE'].nunique()

In [ ]:
df20['OFFENSE_CODE'].value_counts()

So code 3301 leads with 6234 reports in 2020, followed by code 3115, then 801, then 3005, and then 3831. Code 3005 is missing from their list so we have no idea what it is!! That is a crime in itself.   

In [ ]:
# Just another way to accomplish the same thing but in a more complicated way.

df20.groupby('OFFENSE_CODE')['OFFENSE_CODE'].count().reset_index(name='count').sort_values(['count'], ascending = False) 

Let us focus in on these verbal disputes. We will do so by creating a new dataframe that is only for OFFENSE_CODE 3301.

In [ ]:
dfverbal = df20[ df20['OFFENSE_CODE'] == 3301 ]

In [ ]:
# Now we see this dataframe just to check

dfverbal

Which days of the week have more verbal disputes?

In [ ]:
dfverbal['DAY_OF_WEEK'].value_counts()

Which hour, which streets have the most verbal disputes?

In [ ]:
dfverbal['HOUR'].value_counts()

In [ ]:
dfverbal['STREET'].value_counts()

What districts are the worst?

In [ ]:
dfverbal['DISTRICT'].value_counts()

Lookup districts C11, B3, and B2 ... what areas are these?

# Practice Task 01

Pick another data-set from `data.boston.gov` and go through the same commands, picking some interesting element of the dataframe to explore

In [11]:
dfcar = pd.read_csv('https://data.boston.gov/dataset/7b29c1b2-7ec2-4023-8292-c24f5d8f0905/resource/e4bfe397-6bfc-49c5-9367-c879fac7401d/download/tmpajuss2fk.csv')

In [13]:
dfcar.head(5)

dispatch_ts mode_type location_type    street       xstreet1  \
0  2015-01-01 00:24:27+00        mv  Intersection       NaN       TRAIN ST   
1  2015-01-01 03:50:33+00        mv        Street  RIVER ST      WALTER ST   
2  2015-01-01 10:14:13+00       ped  Intersection       NaN      LONDON ST   
3  2015-01-01 18:23:57+00      bike  Intersection       NaN       OLNEY ST   
4  2015-01-01 18:42:19+00       ped  Intersection       NaN  WASHINGTON ST   

       xstreet2     x_cord      y_cord        lat       long  
0   WESTGLOW ST  777243.68  2930930.11  42.289750 -71.052515  
1   WINTHROP ST  758927.71  2918981.60  42.257079 -71.120105  
2   MERIDIAN ST  780725.19  2961410.17  42.373337 -71.039040  
3     INWOOD ST  772710.48  2936614.62  42.305413 -71.069163  
4  COLUMBUS AVE  764813.61  2940364.63  42.315809 -71.098290

In [14]:
dfcar.tail(10)

dispatch_ts mode_type location_type               street  \
23943  2020-08-30 20:49:54+00       ped        Street        BLUE HILL AVE   
23944  2020-08-30 20:50:39+00      bike        Street       HUNTINGTON AVE   
23945  2020-08-30 23:10:01+00        mv        Street        BLUE HILL AVE   
23946  2020-08-30 23:47:24+00        mv  Intersection                  NaN   
23947  2020-08-31 00:46:30+00        mv        Street  AMERICAN LEGION HWY   
23948  2020-08-31 07:42:19+00        mv  Intersection                  NaN   
23949  2020-08-31 12:46:41+00        mv        Street      FOREST HILLS ST   
23950  2020-08-31 17:00:53+00        mv  Intersection                  NaN   
23951  2020-08-31 17:38:23+00       ped  Intersection                  NaN   
23952  2020-08-31 22:30:20+00        mv  Intersection                  NaN   

           xstreet1        xstreet2     x_cord      y_cord        lat  \
23943   HAVELOCK ST     WOODROW AVE  766771.53  2929435.86  42.285763   
23944  CRAFTSON WAY        HEATH ST  761343.16  2945550.95  42.330017   
23945    WILMORE ST    FESSENDEN ST  766363.63  2927058.68  42.279262   
23946   COLUMBIA RD    COLUMBIA TER  771608.67  2938153.63  42.309651   
23947     BRIDGE ST    WALK HILL ST  762618.88  2929938.71  42.287511   
23948       HIGH ST     CONGRESS ST  776546.37  2954582.16  42.354662   
23949   ROSSMORE RD     BROOKLEY RD  763161.11  2935565.21  42.302655   
23950     HENDRY ST      BOWDOIN ST  774005.80  2937832.63  42.308737   
23951   ATKINSON ST  SOUTHAMPTON ST  772685.21  2946272.18  42.331914   
23952  WALK HILL ST       BOURNE ST  760767.71  2932091.36  42.293157   

            long  
23943 -71.091062  
23944 -71.111323  
23945 -71.093052  
23946 -71.073208  
23947 -71.106898  
23948 -71.054636  
23949 -71.104475  
23950 -71.064352  
23951 -71.069076  
23952 -71.113388

In [15]:
dfcar['street'].value_counts()

WASHINGTON ST        657
BLUE HILL AVE        397
DORCHESTER AVE       251
COLUMBIA RD          243
MASSACHUSETTS AVE    235
                    ... 
GOODWIN CT             1
SEAVERNS AVE           1
CHURCH ST              1
GERMANIA ST            1
IROQUOIS ST            1
Name: street, Length: 1259, dtype: int64

In [16]:
dfcar['location_type'].value_counts()

Intersection    11718
Street          10269
Other            1966
Name: location_type, dtype: int64

In [17]:
dfcar['mode_type'].value_counts()

mv      17566
ped      4078
bike     2309
Name: mode_type, dtype: int64

 ## What if we have local data, sitting in a folder called, data?¶

In [ ]:
bm15 = pd.read_csv("data/marathon_results_2015.csv")
bm15.head()

In [ ]:
bm16 = pd.read_csv("data/marathon_results_2016.csv")
bm16.head()

In [ ]:
bm17 = pd.read_csv("data/marathon_results_2017.csv")
bm17.head()

Notice the extra first column in `bm15` and `bm17` ... these should be dropped. 

We can then create an indicator for which year the race data are from, and then combine all three data-frames so that we have a single file. When we do this we will write it out as a `csv` file too.

In [ ]:
bm15 = bm15.drop(columns = 'Unnamed: 0')
bm15.head()

In [ ]:
bm17 = bm17.drop(columns = 'Unnamed: 0')
bm17.head()

In [ ]:
bm15['Year'] = '2015'
bm16['Year'] = '2016'
bm17['Year'] = '2017'

Now we bind the rows (i.e., stack one data-frame on top of the other with `concat( [dfi, dfj, dfk] ) 

In [ ]:
bm_df = pd.concat([bm15, bm16, bm17])

In [ ]:
bm_df.shape 

# How many rows and how many columns do we have? 
# This is a good step to check if we stacked the data-frames correctly or not

### How many runners per year?

In [ ]:
bm_df['Year'].value_counts()

### How many Male/Female runners per year?

In [ ]:
bm_df.groupby('M/F')['Year'].value_counts()

In [ ]:
# The proportion of Male/Female runners, by Year

bm_df.groupby('Year')['M/F'].value_counts(normalize = True)